nixos/modules/services/misc/gitolite.nix

```diff
@@ -77,21 +159,57 @@ in
       serviceConfig.Type = "oneshot";
       serviceConfig.RemainAfterExit = true;
 
-      path = [ pkgs.gitolite pkgs.git pkgs.perl pkgs.bash config.programs.ssh.package ];
-      script = ''
-        cd ${cfg.dataDir}
-        mkdir -p .gitolite/logs
-        if [ ! -d repositories ]; then
-          gitolite setup -pk ${pubkeyFile}
-        fi
-        if [ -n "${hooks}" ]; then
-          cp ${hooks} .gitolite/hooks/common/
-          chmod +x .gitolite/hooks/common/*
-        fi
-        gitolite setup # Upgrade if needed
-      '';
+      path = [ pkgs.gitolite pkgs.git pkgs.perl pkgs.bash pkgs.diffutils config.programs.ssh.package ];
+      script =
+      let
+        rcSetupScriptIfCustomFile =
+          if cfg.manageGitoliteRc then ''
+            cat <<END
+            <4>WARNING: Since NixOS 17.09, .gitolite.rc is managed
+            <4>declaratively by default. You seem to have a manually
+            <4>customized .gitolite.rc file, so it won't be replaced
+            <4>with a NixOS-managed symlink to preserve your changes.
+            <4>See the services.gitolite.manageGitoliteRc option
+            <4>in "man configuration.nix" for more information.
+            END
+          '' else ''
+            :
+          '';
+        rcSetupScriptIfDefaultFileOrStoreSymlink =
+          if cfg.manageGitoliteRc then ''
+            ln -sf "${rcDir}/gitolite.rc" "$GITOLITE_RC"
+          '' else ''
+            rm -f "$GITOLITE_RC"
+          '';
+      in
+        ''
+          cd ${cfg.dataDir}
+          mkdir -p .gitolite/logs
+
+          GITOLITE_RC=.gitolite.rc
+          GITOLITE_RC_DEFAULT=${rcDir}/gitolite.rc.default
+          if ( [[ ! -e "$GITOLITE_RC" ]] && [[ ! -L "$GITOLITE_RC" ]] ) ||
+             ( [[ -f "$GITOLITE_RC" ]] && diff -q "$GITOLITE_RC" "$GITOLITE_RC_DEFAULT" >/dev/null ) ||
+             ( [[ -L "$GITOLITE_RC" ]] && [[ "$(readlink "$GITOLITE_RC")" =~ ^/nix/store/ ]] )
+          then
+        '' + rcSetupScriptIfDefaultFileOrStoreSymlink +
+        ''
```

Comment:
Hm, this mixing of Nix and shell looks more complicate than necessary, I think. There is at least one if/else in Nix+sh that cancel each other out, if you take another look.

I think using this style when expanding nix values inside strings is more readable:

```
  ''hey, a nix value ${var} in the middle of the text''
```

IMHO, the added complexity of checking if .gitolite.rc is unchanged is not worth it. Personally, I'd say "given that manageGitoliteRc is set, is .gitolite.rc a symlink to nix store? If yes, update file, else complain."

If you make some of these changes, all of "rcDir, rcDirScript, rcAppendFile, rcAppendText" can become just "rcFile":

```
rcFile = runCommand "gitolite.rc" {}
  ''
    export HOME=gitolite-home
    mkdir -p gitolite-home/.gitolite/logs
    ${gitolite}/bin/gitolite print-default-rc > "$out"

    cat << EOF >> "$out"

    ${cfg.extraGitoliteRc}
    EOF
  ''
```

Love the documentation!

Reply:
> Hm, this mixing of Nix and shell looks more complicate than necessary, I think. There is at least one if/else in Nix+sh that cancel each other out, if you take another look.

Could you elaborate? I don't see it. Nix and shell conditions operate on different data.

> I think using this style when expanding nix values inside strings is more readable ...

It is surely more readable, but if I put `${rcSetupScriptIfDefaultFileOrStoreSymlink}` on a separate line inside a multi-line string, it will introduce an extra empty line after the contents of the variable, which doesn't look pretty. One could argue that readability of the source is more important that readability of the generated code, though...

> IMHO, the added complexity of checking if .gitolite.rc is unchanged is not worth it.

I'm pretty sure it's important. The thing it, `gitolite setup` creates a default `.gitolite.rc` file if it doesn't exist. So all existing users of gitolite will already have it. And we don't want to bug them with warnings if they didn't actually make any customizations to the file, or make them wonder why their declarative configuration doesn't work, in case they add it and don't notice the warning.

Reply:
> all of "rcDir, rcDirScript, rcAppendFile, rcAppendText" can become just "rcFile":

Embedding `${cfg.extraGitoliteRc}` into the shell script creates a danger of misinterpretation of its contents by the shell (if it has a lone `EOF` line, for some reason). My approach doesn't have this problem.

Reply:
> Could you elaborate? I don't see it. Nix and shell conditions operate on different data.

Perhaps I described it wrong. A concrete example of what I'm thinking of is replacing

```
'' + rcSetupScriptIfDefaultFileOrStoreSymlink +
```
(which has a long name and a if/else in its implementation)

with

```
${lib.optionalString cfg.manageGitoliteRc ''
  echo "warning message"   # or echo "${warningMessage}" if you don't want multi-line shell code here in the .nix file
''}
```

In other words, I find some of the abstractions in this code too abstract. (But yes, that does probably add a blank line in case the condition is false.)

> I'm pretty sure it's important. The thing it, gitolite setup creates a default .gitolite.rc file if it doesn't exist. So all existing users of gitolite will already have it. And we don't want to bug them with warnings if they didn't actually make any customizations to the file, or make them wonder why their declarative configuration doesn't work, in case they add it and don't notice the warning.

Good point. Though I figured now that there is a manageGitoliteRc option the warning message could be turned into an error, and that users perhaps better be made aware of the change from imperative to declarative config from the start. Even if they didn't make a change to the file. But OK, I get your reasoning.

> Embedding ${cfg.extraGitoliteRc} into the shell script creates a danger of misinterpretation of its contents by the shell (if it has a lone EOF line, for some reason). My approach doesn't have this problem.

You can use ```echo '${foo}' > file``` if you don't like using heredoc. (You can also use a more special string than EOF. THIS_ENDS_THE_FILE for example?)

Reply:
> In other words, I find some of the abstractions in this code too abstract. (But yes, that does probably add a blank line in case the condition is false.)

While not ideal, I find that the approach (using named shell script fragments) is the best way (known to me) to convey the intent of the code. Embedding Nix if-then-else inside shell's if-then-else would look messy and confusing. (And we can't use `optionalString` instead of if-then-else, because shell syntax doesn't allow us to leave then- or else- branch empty.)

> ... the warning message could be turned into an error ...

Does this mean "exit with error", or just "write journal message with loglevel 3 instead of 4"?

> You can use echo '${foo}' > file if you don't like using heredoc.

I found a way to properly escape it:

```
      optionalString (cfg.extraGitoliteRc != "") ''
        echo ${escapeShellArg ''

          # Added by NixOS:
          ${removeSuffix "\n" cfg.extraGitoliteRc}

          # per perl rules, this should be the last line in such a file:
          1;
        ''} >>"$out/gitolite.rc"
      '';
```

Does this look better? It's a bit too heavy with nested code-inside-strings, but it's somewhat shorter, and doesn't use additional variables.

Reply:
> Does this mean "exit with error", or just "write journal message with loglevel 3 instead of 4"?

Both :-)

> [...] Does this look better?

IDK. I think both look kind of verbose. But I'm okay with either. (I would probably have written ```printf -- '%s' '${cfg.extraGitoliteRc}' >> "$out"```.)

Reply:
>> Does this mean "exit with error", or just "write journal message with loglevel 3 instead of 4"?

> Both :-)

I don't think we can do the former. It would break existing installations that have customized `.gitolite.rc`, I suppose this should not happen except in truly exceptional circumstances, right?

I don't mind raising the loglevel to provide better visibility, though. Also, maybe it's worth adding a note to the release information for 17.09, to make visibility even better? (I have **no idea** how this works, though. What is good enough to be added to release notes, how to do it, etc.)

Reply:
> I don't think we can do the former. It would break existing installations that have customized .gitolite.rc,

My thinking is that either we error out if .gitolite.rc is customized when upgrading to NixOS 17.09, or the manageGitoliteRc option would have to be turned off by default.

Hm, defaulting manageGitoliteRc to false is an easy way to not cause breakage...

Release notes for NixOS 17.09 is in nixos/doc/manual/release-notes/rl-1709.xml.

Reply:
> Hm, defaulting manageGitoliteRc to false is an easy way to not cause breakage...

This will not fulfill your original intent of making users "aware of the change from imperative to declarative config from the start". :) And it feels less "natural" for NixOS.

Leaving `manageGitoliteRc` defaulting to true would at least issue the warning.

> Release notes for NixOS 17.09 is in nixos/doc/manual/release-notes/rl-1709.xml.

Do I just prepare a PR for them as well? Should it be a separate PR, or it needs to be included in this one?

Reply:
> This will not fulfill your original intent of making users "aware of the change from imperative to declarative config from the start". :)

Well, there is less need to be aware of something if it doesn't happen :-)

> And it feels less "natural" for NixOS.

Agreed. But since it is a breaking change, we could let the transition go over one release cycle instead of introducing breakage right away. So for NixOS 17.09 the manageGitoliteRc option (=false) is introduced (and mentioned in the release notes). Then for NixOS 18.03 the default can be changed.

> Do I just prepare a PR for them as well? Should it be a separate PR, or it needs to be included in this one?

A separate commit in this PR is preferred.